linalg/blockoperator.cpp

```diff
@@ -194,4 +194,140 @@ BlockDiagonalPreconditioner::~BlockDiagonalPreconditioner()
       }
 }
 
+BlockLowerTriangularPreconditioner::BlockLowerTriangularPreconditioner(
+   const Array<int> & offsets)
+   : Solver(offsets.Last()),
+     owns_blocks(0),
+     nRowBlocks(offsets.Size() - 1),
+     nColBlocks(offsets.Size() - 1),
+     row_offsets(0),
+     col_offsets(0),
+     op(nRowBlocks, nRowBlocks)
+{
+   op = static_cast<Operator *>(NULL);
+   row_offsets.MakeRef(offsets);
+   col_offsets.MakeRef(offsets);
+}
+
+BlockLowerTriangularPreconditioner::BlockLowerTriangularPreconditioner(
+   const Array<int> & row_offsets_,
+   const Array<int> & col_offsets_)
+   : Solver(row_offsets_.Last(), col_offsets_.Last()),
+     owns_blocks(0),
+     nRowBlocks(row_offsets_.Size()-1),
+     nColBlocks(col_offsets_.Size()-1),
+     row_offsets(0),
+     col_offsets(0),
+     op(nRowBlocks, nColBlocks)
+{
+   op = static_cast<Operator *>(NULL);
+   row_offsets.MakeRef(row_offsets_);
+   col_offsets.MakeRef(col_offsets_);
+}
+
+void BlockLowerTriangularPreconditioner::SetDiagonalBlock(int iblock,
+                                                          Operator *op)
+{
+   SetBlock(iblock, iblock, op);
+}
+
+void BlockLowerTriangularPreconditioner::SetBlock(int iRow, int iCol,
+                                                  Operator *opt)
+{
+   op(iRow, iCol) = opt;
+
```

Comment:
Should we add a check `iRow >= iCol` to prevent user to add upper-diagonal blocks?